src/bin/josh_sync.rs

```diff
@@ -0,0 +1,143 @@
+    Push {
+        #[clap(long, default_value(DEFAULT_CONFIG_PATH))]
+        config: PathBuf,
+        /// Branch that should be pushed to your remote
+        branch: String,
+        /// Your GitHub usename where the fork is located
+        username: String,
+    },
+}
+
+fn main() -> anyhow::Result<()> {
+    let args = Args::parse();
+    match args.cmd {
+        Command::Init => {
+            let config = JoshConfig {
+                org: "rust-lang".to_string(),
+                repo: "<repository-name>".to_string(),
+                path: "<relative-subtree-path>".to_string(),
+                last_upstream_sha: None,
+            };
+            config
+                .write(Path::new(DEFAULT_CONFIG_PATH))
+                .context("cannot write config")?;
+            println!("Created config file at {DEFAULT_CONFIG_PATH}");
+        }
+        Command::Pull { config } => {
+            let config = load_config(&config)
+                .context("cannot load config. Run the `init` command to initialize it.")?;
+            let josh = get_josh_proxy()?;
+            let sync = GitSync::new(config.clone(), josh);
+            match sync.rustc_pull() {
+                Ok(result) => {
+                    maybe_create_gh_pr(
+                        &config.config.full_repo_name(),
+                        "Rustc pull update",
+                        &result.merge_commit_message,
+                    )?;
+                }
+                Err(RustcPullError::NothingToPull) => {
+                    eprintln!("Nothing to pull");
+                    std::process::exit(2);
+                }
+                Err(RustcPullError::PullFailed(error)) => {
+                    eprintln!("Pull failure: {error:?}");
+                    std::process::exit(1);
+                }
+            }
+        }
+        Command::Push {
+            username,
+            branch,
+            config,
+        } => {
+            let config = load_config(&config)
+                .context("cannot load config. Run the `init` command to initialize it.")?;
+            let josh = get_josh_proxy()?;
+            let sync = GitSync::new(config.clone(), josh);
+            sync.rustc_push(&username, &branch)
+                .context("cannot perform push")?;
+
+            // Open PR with `subtree update` title to silence the `no-merges` triagebot check
+            println!(
+                r#"You can create the rustc PR using the following URL:
+https://github.com/{UPSTREAM_REPO}/compare/{username}:{branch}?quick_pull=1&title={}+subtree+update&body=r?+@ghost"#,
+                config.config.repo
+            );
+        }
+    }
+
+    Ok(())
+}
+
+fn maybe_create_gh_pr(repo: &str, title: &str, description: &str) -> anyhow::Result<bool> {
+    let gh_available = which::which("gh").is_ok();
+    if !gh_available {
+        println!(
+            "Note: if you install the `gh` CLI tool, josh-sync will be able to create the sync PR for you."
+        );
```

Comment:
I'd rather not see this noise every time I run the tool...

Reply:
It's harder to create a URL for `pull`, unless we push the PR for the user (where? to their fork or to the main subtree repo? - we'd need to ask, same as we do for push).

Reply:
IMO it should be pushing to the fork. Pushing directly to the subtree repo pollutes that repo with branches.
But I think I see your point... we can't even know what that remote is called.

Anyway for Miri this usually runs on CI, not on my machine, so automated PR creation is less necessary. However, now I wonder how this code interactively asking for confirmation will work when it runs on CI...